src/mcts/node.cc

```diff
@@ -32,247 +32,151 @@
+  return oss.str();
 }
 
-void Node::Pool::AllocateNewBatch() REQUIRES(allocations_mutex_) {
-  allocations_.emplace_back(std::make_unique<FreeNode[]>(kAllocationSize));
+/////////////////////////////////////////////////////////////////////////
+// EdgeList
+/////////////////////////////////////////////////////////////////////////
 
-  FreeNode* new_nodes = allocations_.back().get();
-  for (int i = 0; i < kAllocationSize; ++i) {
-    FreeNode* n = new_nodes + i;
-    n->next = reserve_list_;
-    reserve_list_ = n;
-  }
+EdgeList::EdgeList(MoveList moves)
+    : edges_(static_cast<Edge*>(operator new(sizeof(Edge) * moves.size()))),
+      size_(moves.size()) {
+  auto* edge = edges_;
+  for (auto move : moves) new (edge++) Edge(move);
 }
 
-void Node::Pool::ReleaseChildren(Node* node) {
-  Node* next = node->child_;
-  // Iterating manually rather than with iterator, as node is released in the
-  // middle and can be taken by other threads, so we have to be careful.
-  while (next) {
-    Node* iter = next;
-    // Getting next after releasing node, as otherwise it can be reallocated
-    // and overwritten.
-    next = next->sibling_;
-    ReleaseSubtree(iter);
-  }
-  node->child_ = nullptr;
+void EdgeList::DestroyList() {
+  if (!edges_) return;
+  for (int i = 0; i < size_; ++i) edges_[i].~Edge();
+  operator delete(static_cast<void*>(edges_));
+  edges_ = nullptr;
 }
 
-void Node::Pool::ReleaseAllChildrenExceptOne(Node* root, Node* subtree) {
-  Node* child = nullptr;
-  Node* next = root->child_;
-  while (next) {
-    Node* iter = next;
-    // Getting next after releasing node, as otherwise it can be reallocated
-    // and overwritten.
-    next = next->sibling_;
-    if (iter == subtree) {
-      child = iter;
-    } else {
-      ReleaseSubtree(iter);
-    }
-  }
-  root->child_ = child;
-  if (child) {
-    child->sibling_ = nullptr;
-  }
+void EdgeList::operator=(EdgeList&& other) noexcept {
+  DestroyList();
+  edges_ = other.edges_;
+  size_ = other.size_;
+  other.edges_ = nullptr;
+  other.size_ = 0;
 }
 
-void Node::Pool::ReleaseSubtree(Node* node) {
-  Mutex::Lock lock(gc_mutex_);
-  subtrees_to_gc_.push_back(node);
-}
-
-Node::Pool gNodePool;
-
 /////////////////////////////////////////////////////////////////////////
 // Node
 /////////////////////////////////////////////////////////////////////////
 
-Node* Node::CreateChild(Move m) {
-  Node* new_node = gNodePool.AllocateNode();
-  new_node->parent_ = this;
-  new_node->sibling_ = child_;
-  new_node->move_ = m;
-  child_ = new_node;
-  return new_node;
+Node* Node::CreateSingleChildNode(Move move) {
+  assert(!edges_);
+  edges_ = EdgeList({move});
+  child_ = std::make_unique<Node>(this, 0);
+  return child_.get();
+}
+
+void Node::CreateEdges(const MoveList& moves) {
+  assert(!edges_);
+  assert(!child_);
+  edges_ = EdgeList(moves);
 }
 
+Node::ConstIterator Node::Edges() const { return {edges_, &child_}; }
+Node::Iterator Node::Edges() { return {edges_, &child_}; }
```

Comment:
Won't this run into the same compilation problem that bit the small-Move PR?

Reply:
No, there problem was that braces initialization is strict with type narrowing:
```cpp
float x = 1.5f;
int y{x}; // Won't compile, float to int may lose data.
```
Here it doesn't happen, and also here I did compile and it compiles. :)